### PR TITLE
Add unit tests and refactor survey retriever

### PR DESCRIPTION
**Pull request by hacodeorg**
# Description
Moving all survey retriever functions into the same class, and make them class methods. Adding unit tests for those functions.

Specifically, I make these changes
1. Remove `SurveyPipelineWorker` inheritance in `DailySurveyRetriever` class.
The main idea behind this inheritance is to wrap a function inside an object that knows how to extract necessary data from its environment (`context` variable), does some processing and then returns results back to the environment. By doing that, we can stack function-objects in any order to create flexible pipelines. However, it comes with the cost of boiler-plate code and it obfuscates what the function really needs. Removing this layer of abstraction makes it easier to reason about the code and turns the function into a pure function.

2. Move 2 retriever functions from `SurveyPipelineHelper` into `DailySurveyRetriever`.

3. Simplify function signatures and update function comments.

4. Add unit tests for all 3 retriever functions. They cover 2 main scenarios, collecting non-empty and empty results.

## Links
- [jira](https://codedotorg.atlassian.net/browse/PLC-584)

## Testing story
- Unit tests
- Manual testing by visiting a AYW survey page (with results) on local machine and verify that everything works correctly.

### CODE REVIEW

File: dashboard/test/lib/pd/survey_pipeline/daily_survey_retriever_test.rb [hunk 1]
```diff
@@ -10,100 +10,106 @@ class DailySurveyRetrieverTest < ActiveSupport::TestCase
     setup_all do
       @workshop_form_ids = [11_000_000_000_000, 11_000_000_000_001]
       @facilitator_form_ids = [22_000_000_000_000, 22_000_000_000_001]
-      @workshops = create_list :csf_deep_dive_workshop, 2, num_sessions: 2
 
+      @workshops = create_list :csf_deep_dive_workshop, 2, num_sessions: 2
```
Comment (hacodeorg) [inline dashboard/test/lib/pd/survey_pipeline/daily_survey_retriever_test.rb:14]: The number of session is not necessary and can be removed

File: dashboard/lib/pd/survey_pipeline/daily_survey_retriever.rb [hunk 2]
```diff
@@ -1,62 +1,64 @@
...
+    end
 
-      # Collect survey questions; ignore forms that don't have submissions.
-      form_ids_with_submissions =
-        ws_submissions.pluck(:form_id).uniq | facilitator_submissions.pluck(:form_id).uniq
+    # Retrieve general workshop survey submissions and questions for selected workshops.
+    #
+    # @param workshop_ids [Integer, Array<Integer>] a workshop id or a list of workshop ids
+    # @return [Array<Array>] An array of 2 arrays: Array<Pd::SurveyQuestion> and
+    #   Array<Pd::WorkshopDailySurvey>. Returns nil if the input parameter is nil.
```
Comment (islemaster) [inline dashboard/lib/pd/survey_pipeline/daily_survey_retriever.rb]: Note: The `questions` and `submissions` inner return values from this function are not `Array`, they're [`ActiveRecord::Relation`](https://api.rubyonrails.org/classes/ActiveRecord/Relation.html).  If you don't want to expose that that level of detail to the consumer, you could just document them as `Enumerable`.

The same is true some other places in this file.

**Review by islemaster — APPROVED** [review]
This looks great!

It's very interesting to be moving away again from a more generalized implementation to something domain-specific and therefore easier to understand.  I like where you've ended up.

I've got one note on the doc-comments, nonblocking.

File: dashboard/lib/pd/survey_pipeline/daily_survey_retriever.rb (hunk 2 above)
Comment (hacodeorg) [reply dashboard/lib/pd/survey_pipeline/daily_survey_retriever.rb]: Yeah that's more accurate. Will fix.

**Comment by hacodeorg** [comment]
Looking back, the generalized approach is an example of premature optimization. The return is not worth the investment. I should practice ["You Aren't Gonna Need It"](https://www.martinfowler.com/bliki/Yagni.html) principle more.

> This looks great!
> 
> It's very interesting to be moving away again from a more generalized implementation to something domain-specific and therefore easier to understand. I like where you've ended up.
> 
> I've got one note on the doc-comments, nonblocking.

**Pull request merged**